### PR TITLE
feat: Suggest closest command on typo

### LINKED ISSUE #3856: BUG: typos in subcommand lead to confusing warnings
### Troubleshooting docs

- [x] My problem is not solved in the Troubleshooting docs

### Anaconda default channels

- [x] I do NOT use the Anaconda default channels (pkgs/* etc.)

### How did you install Mamba?

Mambaforge or latest Miniforge

### Search tried in issue tracker

repoquery; CLI; parsing

### Latest version of Mamba

- [x] My problem is not solved with the latest version

### Tried in Conda?

Not applicable

### Describe your issue

I've seen CLI parsing weirdness here or there post-2.0, but `repoquery whoneeds` is just straight not usable anymore
```
>mamba reqoquery whoneeds _openmp_mutex
The following arguments were not expected: _openmp_mutex whoneeds reqoquery
Run with --help for more information.
```
(note the inverted arguments in the error message)

```
>mamba reqoquery whoneeds -c conda-forge --platform linux-64 _openmp_mutex
The following arguments were not expected: _openmp_mutex linux-64 --platform conda-forge -c whoneeds reqoquery
Run with --help for more information.
```

```
>mamba reqoquery whoneeds _openmp_mutex -c conda-forge --platform linux-64
The following arguments were not expected: linux-64 --platform conda-forge -c _openmp_mutex whoneeds reqoquery
Run with --help for more information.
```

The same commands work on linux

### mamba info / micromamba info

```shell
>mamba info

       libmamba version : 2.0.5
          mamba version : 2.0.5
           curl version : libcurl/8.11.1 Schannel zlib/1.3.1 libssh2/1.11.1
     libarchive version : libarchive 3.7.7 zlib/1.3.1 liblzma/5.6.3 bz2lib/1.0.8 liblz4/1.10.0 libzstd/1.5.6
       envs directories : E:\miniforge\envs
          package cache : E:\miniforge\pkgs
                          C:\Users\[...]\.mamba\pkgs
                          C:\Users\[...]\AppData\Roaming\.mamba\pkgs
            environment : builder (active)
           env location : E:\miniforge\envs\builder
      user config files : C:\Users\[...]\.mambarc
 populated config files : C:\Users\[...]\.condarc
                          E:\miniforge\.condarc
       virtual packages : __win=10.0.26100=0
                          __archspec=1=x86_64
               channels : https://conda.anaconda.org/conda-forge/noarch
                          https://conda.anaconda.org/conda-forge/win-64
       base environment : E:\miniforge
               platform : win-64
```

### Logs

```shell

```

### environment.yml

```yaml

```

### ~/.condarc

```yaml

```

### PR DESCRIPTION
**Pull request by faze-geek**
# Description

Fixes #3856 This PR adds hints when typos are made to subcommands.

Before : Mamba prints only a cryptic CLI11 error message with no hints.
```
anurag@Anurags-MacBook-Air mamba % mamba instal xtensor
The following arguments were not expected: instal xtensor
Run with --help for more information.
```

Now : Conda style "Did you mean `<cmd>`" suggestion is added.
```
anurag@Anurags-MacBook-Air mamba % mamba instal xtensor
The following arguments were not expected: instal xtensor
Did you mean 'install'?
Run with --help for more information.
```

## Type of Change

- [ ] Bugfix
- [x] Feature / enhancement
- [ ] CI / Documentation
- [ ] Maintenance

## Checklist

- [x] My code follows the general style and conventions of the codebase, ensuring consistency
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have run `pre-commit run --all` locally in the source folder and confirmed that there are no linter errors.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

**Comment by faze-geek** [comment]
Approach

- Added two utilities to `libmamba/util`
   1. `similarity_ratio(str1, str2)` : Mimics Python's [difflib.SequenceMatcher(str1, str2).ratio()](https://docs.python.org/3/library/difflib.html#difflib.SequenceMatcher.ratio).
   2. `closest_matches(input, candidates, cutoff, max_results)` : Mimics [difflib.get_close_matches](https://docs.python.org/3/library/difflib.html#difflib.get_close_matches). 
   It ensures conda parity, as it uses the same [algorithm](https://github.com/conda/conda/blob/main/conda/exceptions.py).

- Hooked into CLI11 via `App::failure_message` : Descent to the deepest parsed subcommand, take the first non flag leftover token & return the single best match.

**Comment by faze-geek** [comment]
More testing from my terminal

**1. Top level typo**
```
anurag@Anurags-MacBook-Air mamba % mamba instal xtensor
The following arguments were not expected: instal xtensor
Did you mean 'install'?
Run with --help for more information.
```
**2. Transition typo**
```
anurag@Anurags-MacBook-Air mamba % $MAMBA lsit
The following argument was not expected: lsit
Did you mean 'list'?
Run with --help for more information.
anurag@Anurags-MacBook-Air mamba % $MAMBA activ
```
**3. Truncated**
```
The following argument was not expected: activ
Did you mean 'activate'?
Run with --help for more information.
```
**4. Subcommand**
```
anurag@Anurags-MacBook-Air mamba % $MAMBA repoquery whoneds openmp
The following arguments were not expected: whoneds openmp
Did you mean 'whoneeds'?
Run with --help for more information.
```
**5. Alias**
```
anurag@Anurags-MacBook-Air mamba % $MAMBA uninstal xtensor
The following arguments were not expected: uninstal xtensor
Did you mean 'uninstall'?
Run with --help for more information.
```
**6. Flags do not get suggestions**
```
anurag@Anurags-MacBook-Air mamba % $MAMBA list --badflag
The following argument was not expected: --badflag
Run with --help for more information.
```
**7. Randoms do not get suggestions**
```
anurag@Anurags-MacBook-Air mamba % $MAMBA zzzzzzzz
The following argument was not expected: zzzzzzzz
Run with --help for more information.
```

**Comment by jjerphan** [comment]
Hello @faze-geek,

Thank you for this contribution.

Although I also think it makes sense for mamba to suggest commands, I wonder whether this issue could better be handled by CLI11 as proposed by @Hind-M with https://github.com/CLIUtils/CLI11/issues/1149.

Edit: I guess we could propose this implementation to CLI11 and in the meantime vendor it in mamba.

**Comment by faze-geek** [comment]
Hi @jjerphan, thanks for the background.

I did go through https://github.com/CLIUtils/CLI11/issues/1149 and see 2 distinct things -

1. Prefix Matching - This makes `myapp ins` run `install` when the prefix is unambiguous. Auto execution is a different functionality that the typo feature requested in #3856 .
2. Close Matching - Not in the library, lives in an `examples/` file for exploration. @phlptp does explicitly mention in comments that they are not ready to integrate close matching as a standard feature (or whether or not that should be included at all). 

Also I feel the implementation here `ratcliff_obershelp_matches` is polished for conda parity compared to plain `Levenshtein Distance` in there (mishandles typos like `lsit`, `ifno` etc.)

The design here already anticipates your vendor solution as it's essentially two pure, dependency-free, unit tested functions isolated in `libmamba/util`. We delete them and swap the hook if/when CLI11 promotes close matching.

Happy to open a CLI11 issue proposing this as a feature and keep working on the mamba fix as well ! Would like to know what you think ?

**Comment by jjerphan** [comment]
> Happy to open a CLI11 issue proposing this as a feature and keep working on the mamba fix as well ! Would like to know what you think ?

+1 for mentioning this solution in https://github.com/CLIUtils/CLI11/issues/1149 and having this implementation in this code-base for now.

File: libmamba/tests/src/util/test_string.cpp [hunk 1]
```diff
@@ -605,5 +605,170 @@ namespace
...
+                    REQUIRE(ratio <= previous + 1e-9);
+                    previous = ratio;
+                }
+            }
+
+            SECTION("max_results caps the number of suggestions")
+            {
+                // A permissive cutoff matches several commands; cap to 2.
+                const auto matches = closest_matches("in", commands, 0.1, 2);
+                REQUIRE(matches.size() <= 2);
```
Comment (jjerphan) [inline libmamba/tests/src/util/test_string.cpp]:
```suggestion
                // A permissive cutoff matches several commands; cap to 2.
                const auto matches = closest_matches("in", commands, 0.1, 2);
                REQUIRE(matches.size() > 2);

                const auto matches_capped = closest_matches("in", commands, 0.1, 2);
                REQUIRE(matches_capped.size() <= 2);
```

File: libmamba/tests/src/util/test_string.cpp [hunk 2]
```diff
@@ -605,5 +605,170 @@ namespace
...
+                const auto matches = closest_matches("instal", commands, 0.6, 5);
+                REQUIRE(matches.size() >= 2);
+
+                double previous = 1.0;
+                for (const auto& m : matches)
+                {
+                    const double ratio = similarity_ratio(m, "instal");
+                    REQUIRE(ratio <= previous + 1e-9);
+                    previous = ratio;
+                }
```
Comment (jjerphan) [inline libmamba/tests/src/util/test_string.cpp]: Can you use [`std::is_sorted`](https://en.cppreference.com/cpp/algorithm/is_sorted) here?


```suggestion
                REQUIRE(std::is_sorted(matches, std::greater<double>));
```

File: libmamba/tests/src/util/test_string.cpp [hunk 3]
```diff
@@ -605,5 +605,170 @@ namespace
...
+
+            SECTION("Truncated prefixes are matched")
+            {
+                // "active" -> "activate" has ratio 0.77>= 0.6.
+                REQUIRE(closest_matches("active", commands).front() == "activate");
+            }
+
+            SECTION("repoquery -> repoquery")
+            {
+                const auto matches = closest_matches("Repoquery", commands);
```
Comment (jjerphan) [inline libmamba/tests/src/util/test_string.cpp:703]: Can we add more test cases for cases with upper case letter to specify edge-cases?

E.g. Should the similarity ratio of "Repoquery" and "repoquery" be one?

File: libmamba/tests/src/util/test_string.cpp [hunk 4]
```diff
@@ -605,5 +605,170 @@ namespace
...
+            }
+
+            SECTION("Case Sensitivity: comparison is byte-exact")
+            {
+                REQUIRE(similarity_ratio("List", "list") == Catch::Approx(0.75));
+            }
+
+            SECTION("Ratio is bounded between 0 and 1")
+            {
+                const char* words[] = { "install", "lst", "repoquery", "", "x", "activate" };
```
Comment (jjerphan) [inline libmamba/tests/src/util/test_string.cpp]: Prefer:


```suggestion
                constexpr std::array<std::string_view, 6> words = { "install", "lst", "repoquery", "", "x", "activate" };
```

File: libmamba/tests/src/util/test_string.cpp [hunk 5]
```diff
@@ -605,5 +605,170 @@ namespace
...
+                        REQUIRE(ratio >= 0.0);
+                        REQUIRE(ratio <= 1.0);
+                    }
+                }
+            }
+        }
+
+        TEST_CASE("closest_matches")
+        {
+            const std::vector<std::string> commands = {
```
Comment (jjerphan) [inline libmamba/tests/src/util/test_string.cpp:672]:
```suggestion
            constexpr std::array<std::string_view, 17> commands = {
```

File: libmamba/src/util/string.cpp [hunk 6]
```diff
@@ -989,4 +989,125 @@ namespace mamba::util
...
+        std::vector<Scored> scored;
+        scored.reserve(candidates.size());
+        for (const auto& candidate : candidates)
+        {
+            const double ratio = similarity_ratio(input, candidate);
+            if (ratio >= cutoff)
+            {
+                scored.push_back({ ratio, &candidate });
+            }
+        }
```
Comment (jjerphan) [inline libmamba/src/util/string.cpp]: Neatpick: We could make use of C++20 here.
```cpp
namespace views = std::ranges::views;

auto scored_view = candidates
                   | views::transform(
                         [&](const std::string& candidate) -> Scored
                         { return { similarity_ratio(input, candidate), &candidate }; }
                     )
                   | views::filter([cutoff](const Scored& s) { return s.ratio >= cutoff; });

// TODO(C++23): std::ranges::to<std::vector>
auto scored = std::vector<Scored>(scored_view.begin(), scored_view.end());
```

File: libmamba/src/util/string.cpp [hunk 7]
```diff
@@ -989,4 +989,125 @@ namespace mamba::util
...
+        );
+
+        std::vector<std::string> results;
+        const std::size_t count = std::min(max_results, scored.size());
+        results.reserve(count);
+        for (std::size_t i = 0; i < count; ++i)
+        {
+            results.push_back(*scored[i].value);
+        }
+        return results;
```
Comment (jjerphan) [inline libmamba/src/util/string.cpp]: Also neatpick.
```suggestion
        auto results_view = scored
                    | views::take(max_results)
                    | views::transform([](const Scored& s) { return *s.value; });

        // TODO(C++23): std::ranges::to<std::vector>
        return std::vector<std::string>(results_view.begin(), results_view.end());
```

File: libmamba/src/util/string.cpp [hunk 8]
```diff
@@ -989,4 +989,125 @@ namespace mamba::util
...
+            std::vector<std::size_t> j2len(b.size(), 0);
+            for (std::size_t i = 0; i < a.size(); ++i)
+            {
+                std::vector<std::size_t> new_j2len(b.size(), 0);
+                for (std::size_t j = 0; j < b.size(); ++j)
+                {
+                    if (a[i] != b[j])
+                    {
+                        continue;
+                    }
```
Comment (jjerphan) [inline libmamba/src/util/string.cpp:1024]: Should we relax this test to be case insensitive?

**Review by jjerphan — APPROVED** [review]
LGTM.

Mostly nitpicks.

f24b6286b7becedf3e7d0ea852702c0a9458ae3e aggregates some of the suggestions.

File: libmamba/tests/src/util/test_string.cpp (hunk 1 above)
Comment (faze-geek) [reply libmamba/tests/src/util/test_string.cpp]: Good idea to also assert the uncapped case is larger - the first call needs a large `max_results` (like 100), since with `max_results` = 2, the `size() > 2` assertion can never hold.

File: libmamba/tests/src/util/test_string.cpp (hunk 2 above)
Comment (faze-geek) [reply libmamba/tests/src/util/test_string.cpp]: Sure, I shall adopt `std::is_sorted` !

File: libmamba/tests/src/util/test_string.cpp (hunk 3 above)
Comment (faze-geek) [reply libmamba/tests/src/util/test_string.cpp:703]: Yes agreed, I'm adding a set of explicit upper test cases so the byte exact behavior is documented !

Comment (faze-geek) [reply libmamba/tests/src/util/test_string.cpp:703]: F.Y.I, the implementation is case-sensitive & that matches conda.

The similarity will deliberately capture the leading `R` / `r` difference and result into `2*8/18 ~ 0.888889`.

File: libmamba/src/util/string.cpp (hunk 8 above)
Comment (faze-geek) [reply libmamba/src/util/string.cpp:1024]: I'd push back on that. I've added explicit upper case tests to show the helper works as expected !

**Comment by faze-geek** [comment]
Thank you so much @jjerphan for going through this. Have addressed most things (& resolved as well as I agree with everything other than the case senstivity pointer), waiting for the tests to run on the latest !

**Comment by faze-geek** [comment]
@jjerphan Do you see anything left to tackle here ?

**Comment by jjerphan** [comment]
There are a few failures left in the tests suite.

**Comment by faze-geek** [comment]
Sorry @jjerphan, it took a while. I couldn't pinpoint those failures as the pre commit hooks changed my diff right before the commit. Realized it now and should be ready to go in. Thanks !

**Comment by jjerphan** [comment]
Thank you, @faze-geek!
